src/main/java/build/buildfarm/instance/shard/ShardInstance.java

```diff
@@ -1195,21 +1197,22 @@ boolean next() {
         () ->
             notFoundNull(
                 expect(directoryBlobDigest, Directory.parser(), executor, requestMetadata));
-    // is there a better interface to use for the cache with these nice futures?
+
+    BiFunction<Digest, Executor, CompletableFuture<Directory>> getCallback =
+        new BiFunction<Digest, Executor, CompletableFuture<Directory>>() {
+          @Override
+          public CompletableFuture<Directory> apply(Digest digest, Executor executor) {
+            logger.log(
+                Level.FINE,
+                format(
+                    "transformQueuedOperation(%s): fetching directory %s",
+                    reason, DigestUtil.toString(directoryBlobDigest)));
+            return toCompletableFuture(fetcher.get());
+          }
+        };
+
     return catching(
-        directoryCache.get(
-            directoryBlobDigest,
-            new Callable<ListenableFuture<? extends Directory>>() {
-              @Override
-              public ListenableFuture<Directory> call() {
-                logger.log(
-                    Level.FINE,
-                    format(
-                        "transformQueuedOperation(%s): fetching directory %s",
-                        reason, DigestUtil.toString(directoryBlobDigest)));
-                return fetcher.get();
-              }
-            }),
+        toListenableFuture(directoryCache.get(directoryBlobDigest, getCallback)),
```

Comment:
I don't think this catching is necessary at this point. As indicated by @ben-manes, I believe that a missing entry from caffeine will return a null. Please verify with tests and ensure we've got coverage.

Reply:
fixed